content/posts/2018-09-09-optimizing-vectors.md

```diff
@@ -0,0 +1,257 @@
+is the same as `L4,4 v2` from Android Studio — it draws a line to `4, 4` and another one to `4, 6`.
+`L` draws a line to the point and `v` draws a vertical line using the shift, which is `6 - 4 = 2` in our case.
+
+Operations squashing helps with optimizing drawing performance. Fewer commands
+mean more efficient execution. Actually, Android Lint
+[has a `VectorPath` check](http://tools.android.com/tips/lint-checks)
+for such cases.
+
+> :book: The maximum instructions count for Lint is `800`.
+> See [the source code](https://android.googlesource.com/platform/tools/base/+/studio-master-dev/lint/libs/lint-checks/src/main/java/com/android/tools/lint/checks/VectorPathDetector.java)
+> for details.
+
+Besides `pathData` differences, there are some minor deviations as well.
+
+* `svg2android` tends to keep useless `<path>` — ones that draw nothing, i. e. transparent lines.
+  Other tools remove them.
+* `svg2vector` tends to keep `<group>` containers.
+  Other tools merge it with inner `<path>` when possible.
+
+## Performance
+
+### Method
+
+Most of the time `VectorDrawable` goes straight into `ImageView`.
+Restarting application over and over to get necessary numbers
+is kind of tedious, especially if there are going to be thousands of passes.
+The benchmark code measures `VectorDrawable#draw` calls instead.
+
+<details>
+  <summary>_Click to expand benchmark code._</summary>
+
+```kotlin
+companion object {
+    private const val ITERATIONS_COUNT = 30_000
+}
+
+private val random = Random()
+
+private fun measureRenderMillis() {
+    listOf(
+            R.drawable.android_studio,
+            R.drawable.svg2android,
+            R.drawable.svg2vector,
+            R.drawable.android_studio_and_avocado,
+            R.drawable.svgo_and_android_studio
+    ).forEach { res ->
+
+        val name = resources.getResourceEntryName(res)
+
+        val averageRenderMillis = (0..ITERATIONS_COUNT)
+            .map { measureRenderMillis(res) }
+            .average()
+
+        println(":: [$name]: $averageRenderMillis ms")
+    }
+}
+
+private fun measureRenderMillis(@DrawableRes drawableRes: Int): Long {
+    val drawable = getDrawable(drawableRes).apply {
+        if (this !is VectorDrawable) {
+            throw IllegalArgumentException("Drawable supposed to be VectorDrawable.")
+        }
+
+        // Tint to prevent internal caching, i. e. draw from scratch all the time.
+        setTint(random.nextInt())
+    }
+
+    val bitmap = Bitmap.createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight, Bitmap.Config.ARGB_8888)
+    val canvas = Canvas(bitmap)
+
+    drawable.setBounds(0, 0, canvas.width, canvas.height)
+
+    return measureTimeMillis { drawable.draw(canvas) }
+}
+```
+
+</details>
+
+### Environment
+
+* MacBook Pro: Intel Core i5 @ 2.7 GHz, 8 GB 1867 MHz DDR3.
+* Emulator: Nexus 4, API 26.
+* Iterations count: 30 000.
+
+### Results
+
+#### `octicon-octoface`
+
+Tool                     | Average Duration, ms
```

Comment:
Please change 2nd column title to something like `Average Vector Drawing Duration, ms`

Otherwise table reads like: it takes n ms to do something (process/optimize) using one of the tools, not to draw them on Android